src/bootstrap/src/core/builder/tests.rs

```diff
@@ -189,6 +189,53 @@ fn alias_and_path_for_library() {
     assert_eq!(first(cache.all::<doc::Std>()), &[doc_std!(A => A, stage = 0)]);
 }
 
+#[test]
+fn ci_rustc_if_unchanged_logic() {
```

Comment:
This checks if `rust.download-rustc='if-unchanged'` works correctly in dry-run mode (which means this is very cheap to run in any runner) depending on whether the compiler has been modified (e.g., when the compiler tree is not modified, the test expects bootstrap to choice ci rustc).

There's an edge case: if a PR changes something in a way that breaks the logic of building compiler when it's changed but doesn't modify the compiler code, our current test won't detect this issue. And that can lead us to unintentionally merge this bug to upstream. So I made `mingw-check` runner to temporarily modify the compiler (https://github.com/rust-lang/rust/pull/122709/commits/3665a98976817e074475a27089c6c7252abeb69a) and execute this test to cover this edge case.

@rustbot ready